### ISSUE TITLE
proj_trans returning the incorrect value for ellipsoid to geoid transform

### ISSUE THREAD

**Issue by AspenEyers** [OSGeo/PROJ#2566]
Hi, 
I'm not sure if this is intentional or not (I'm pretty new to proj), but it seems like an issue.
I am trying to do an altitude conversion to determine the height offset between a MSL geoid model EGM2008 and the WGS84 ellipsoid.

I create a `proj_create_crs_to_crs` between the two models which are `EPSG:4326` and `EPSG:3855`. 
But when I then put in any lat,long,alt value, the returned answer is the same lat,long,alt but the altitude is negative. 

The same situation can be seen in the following link, but instead, it is  using pyproj https://gis.stackexchange.com/questions/366769/adding-epsg-codes-to-pyproj. 

After a wonderfully informative google introduction into all things altitude-related I realized that I might not have the dataset downloaded or in my path. As it turns out this was the case, I needed to download the model `egm08_25.gtx_tmp` and I added it to my `$HOME/.local/share/proj` which fixed my issue. 

A short example of the code I was running:
```
    double in_lat = -33.80;
    double in_lon = 151.15;
    double in_alt = 60;
    double out_alt = 82.59;

    C = proj_context_create();
    P = proj_create_crs_to_crs (C, "EPSG:4326", "EPSG:3855", NULL);
    PJ_COORD WGS84_point = proj_coord (in_lat, in_lon, in_alt, 0);
    PJ_COORD MSL_point = proj_trans (P, PJ_INV, WGS84_point);`
    
    std::cerr <<"FROM - u: " <<  WGS84_point.uvw.u << " v: " <<  WGS84_point.uvw.v << " w: " <<  WGS84_point.uvw.w << std::endl;
    std::cerr << "TO   - u: "  <<  MSL_point.uvw.u << " v: " <<  MSL_point.uvw.v << " w: " <<  MSL_point.uvw.w << std::endl;
```
Before I added the dataset to my path the output I got was:
```
FROM - u:  -33.80 v:   151.15 w:  60
TO      - u:  -33.80  v:   151.15 w: -60
```
And afterwards It was changed to correctly be: (according to some online calculators)
```
FROM - u:  -33.80 v:   151.15 w:  60
TO   - u:  -33.80    v:   151.15 w:   82.59
```

My guess is that when there is no dataset, the transform implicitly becomes 2D or something so the altitude measurement doesn't make much sense? I'm really not sure here but it does seem like an issue...

**Comment by kbevers** [comment]
> I'm really not sure here but it does seem like an issue

It seems you fixed it all by yourself. Without the grid we can't do anything sensible with your vertical component. I am not entirely sure why it is returned with the reverse sign though.

**Comment by AspenEyers** [comment]
Yeah, I guess the reason that I flagged it as a 'bug' is because on the users side it is plausible to get sensible numbers back (or seem to get sensible numbers back) when running without a dataset. 
The error checking then becomes a process of checking to see if files exist within the proj path before running the command and hoping the file has loaded in properly.

I'm hoping to use this code on a drone so a negative sign on the altitude could cause a lot of harm :/

Would it be best if it returned a null when defaulting to a 2d transform (if that is what it is doing) and is there any way I can know that it has performed a 2d instead of 3d transform for more robust error checking?

**Comment by rouault** [comment]
The issue with the sign inversion was fixed per 87270ae6f1840872d5f49c70258941197b0ce809. 
However, geodetically speaking, you can't transform between a 2D CRS (EPSG:4326) and a vertical CRS (EPSG:3855).
You should rather transform between a geographic 3D CRS "EPSG:4979" and a compound CRS "EPSG:4326+3855"
If you want to have full control of what operations are used, if they contain ballpark operations, etc., you should use the lower level API : proj_create_operations() and its related functions (proj_coordoperation_has_ballpark_transformation())

**Comment by AspenEyers** [comment]
> The issue with the sign inversion was fixed per 87270ae.

It still seems odd to me that the altitude should return anything at all be it -'ve or +'ve. To me it seems like returning an altitude in the absence of a .gtx file is misleading when EPSG:3855 is clearly asking for an altitude conversion with a specific file. 
Am I missing something here? 

> "EPSG:4979" and a compound CRS "EPSG:4326+3855" 

Thanks for that, yeah I can see why that's a much better option.

> you should use the lower level API : proj_create_operations() and its related functions 
(proj_coordoperation_has_ballpark_transformation())

I'll look into that thanks!
